From dc4242bd2ebd2d80544d903fb612c1cf153e5497 Mon Sep 17 00:00:00 2001
From: Richard Laager <rlaager@wiktel.com>
Date: Sat, 30 May 2020 20:39:31 -0500
Subject: [PATCH 2/2] Fix another dependency loop

zfs-load-key-DATASET.service was gaining an
After=systemd-journald.socket due to its stdout/stderr going to the
journal (which is the default).  systemd-journald.socket has an After
(via RequiresMountsFor=/run/systemd/journal) on -.mount.  If the root
filesystem is encrypted, -.mount gets an After
zfs-load-key-DATASET.service.

By setting stdout and stderr to null on the key load services, we avoid
this loop.

Signed-off-by: Richard Laager <rlaager@wiktel.com>
Closes: #10356
---
 etc/systemd/system-generators/zfs-mount-generator.in | 4 ++++
 1 file changed, 4 insertions(+)

--- a/etc/systemd/system-generators/zfs-mount-generator.in
+++ b/etc/systemd/system-generators/zfs-mount-generator.in
@@ -218,6 +218,10 @@
 [Service]
 Type=oneshot
 RemainAfterExit=yes
+# This avoids a dependency loop involving systemd-journald.socket if this
+# dataset is a parent of the root filesystem.
+StandardOutput=null
+StandardError=null
 ExecStart=${keyloadcmd}
 ExecStop=${keyunloadcmd}"   > "${dest_norm}/${keyloadunit}"
     fi
